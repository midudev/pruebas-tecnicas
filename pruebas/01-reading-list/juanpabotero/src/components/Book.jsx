import { useParams } from 'react-router-dom';

export function Book({ books }) {
  const { id } = useParams();
  const book = books.find((book) => String(book.id) === id);
  return (
    <article className="flex flex-col gap-5 items-center border rounded-lg shadow md:flex-row md:max-w-2xl border-gray-700 bg-gray-800 mx-auto p-4">
      <img
        className="object-contain w-full rounded h-96 md:h-auto md:w-48"
        src={book.cover}
        alt={book.title}
      />
      <div className="flex flex-col justify-between leading-normal">
        <h5 className="mb-1 text-2xl md:text-3xl font-bold tracking-tight text-white">
          {book.title}
        </h5>
        <p className="mb-6 text-gray-400">
          <span className="font-bold">Por:</span> {book.author.name}
        </p>
        <p className="mb-6 text-white text-lg leading-snug">{book.synopsis}</p>
        <p className="text-gray-400">
          <span className="font-bold">Género:</span> {book.genre}
        </p>
        <p className="text-gray-400">
          <span className="font-bold">No. de páginas:</span> {book.pages}
        </p>
        <p className="mb-6 text-gray-400">
          <span className="font-bold">Año:</span> {book.year}
        </p>
        <p className="text-gray-400">
          <span className="font-bold">Más libros de {book.author.name}:</span>
        </p>
        <p className="text-gray-400">
          {book.author.otherBooks.length === 0 ? (
            <span key={book.id}>No hay más libros disponibles del autor</span>
          ) : (
            book.author.otherBooks.map((otherBook, index) => {
              return <span key={index}>{otherBook}, </span>;
            })
          )}
        </p>
      </div>
    </article>
  );
}
