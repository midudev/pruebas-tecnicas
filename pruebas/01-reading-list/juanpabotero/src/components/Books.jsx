import { useReadingList } from '../hooks/useReadingList';
import {
  AddBookIcon,
  BookAddedIcon,
  RemoveBookIcon,
  SadFaceIcon
} from './Icons';
import '../styles/Books.css';
import { Link } from 'react-router-dom';

export function Books({ books, isReadingListActive }) {
  const { addToReadingList, removeFromReadingList, readingList } =
    useReadingList();

  const checkBookInReadingList = (product) => {
    return readingList.some((book) => book.id === product.id);
  };

  return (
    <>
      {books.length === 0 && (
        <div className="grid gap-2 justify-items-center text-gray-100 mt-10">
          <SadFaceIcon />
          {isReadingListActive ? (
            <p className="text-center">
              AÃºn no tienes libros en tu lista. <br /> Prueba agregando uno!
            </p>
          ) : (
            <p className="text-center">
              No hay libros disponibles. <br /> Prueba cambiando los filtros.
            </p>
          )}
        </div>
      )}
      <section className="grid gap-8 grid-cols-2 sm:grid-cols-3 md:grid-cols-4 lg:grid-cols-5">
        {books.map((book) => {
          const isBookInReadingList = checkBookInReadingList(book);

          return (
            <article
              key={book.id}
              className="grid relative content-start max-w-[160px] sm:max-w-[178px]
              border rounded-lg shadow bg-gray-800 border-gray-700"
            >
              <button
                className={`add-book-button absolute -top-2 -right-1 rounded-full p-1 bg-gray-900 
              hover:scale-110 transition-all duration-200 z-20
              ${
                isReadingListActive
                  ? 'text-red-600'
                  : isBookInReadingList
                  ? 'text-lime-400'
                  : 'text-yellow-400'
              }`}
                aria-label="addBookButton"
                onClick={() => {
                  isBookInReadingList
                    ? removeFromReadingList(book)
                    : addToReadingList(book);
                }}
              >
                {isReadingListActive ? (
                  <RemoveBookIcon />
                ) : isBookInReadingList ? (
                  <BookAddedIcon />
                ) : (
                  <AddBookIcon />
                )}
              </button>
              <Link to={`/book/${book.id}`}>
                <img
                  className="book-cover h-56 sm:h-64 w-full aspect-auto mb-2 rounded-md"
                  src={book.cover}
                  alt={`Portada de ${book.title}`}
                />
              </Link>
              <div className="p-2 sm:p-3">
                <Link to={`/book/${book.id}`}>
                  <h2 className="text-lg font-bold text-white mb-1">
                    {book.title}
                  </h2>
                </Link>
                <p className="text-gray-400">{book.author.name}</p>
                <p className="text-gray-400 text-sm">{book.genre}</p>
              </div>
            </article>
          );
        })}
      </section>
    </>
  );
}
