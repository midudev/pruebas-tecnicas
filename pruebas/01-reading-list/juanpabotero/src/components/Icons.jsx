export function ReadingListIcon() {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width="32"
      height="32"
      viewBox="0 0 24 24"
    >
      <path
        fill="currentColor"
        d="M17.5 14.33c.79 0 1.63.08 2.5.24v1.5c-.62-.16-1.46-.24-2.5-.24c-1.9 0-3.39.33-4.5.99v-1.69c1.17-.53 2.67-.8 4.5-.8M13 12.46c1.29-.53 2.79-.79 4.5-.79c.79 0 1.63.07 2.5.23v1.5c-.62-.16-1.46-.24-2.5-.24c-1.9 0-3.39.34-4.5.99m4.5-3.65c-1.9 0-3.39.32-4.5 1V9.84c1.23-.56 2.73-.84 4.5-.84c.79 0 1.63.08 2.5.23v1.55c-.74-.19-1.59-.28-2.5-.28m3.5 8V7c-1.04-.33-2.21-.5-3.5-.5c-2.05 0-3.88.5-5.5 1.5v11.5c1.62-1 3.45-1.5 5.5-1.5c1.19 0 2.36.16 3.5.5m-3.5-14c2.35 0 4.19.5 5.5 1.5v14.56c0 .12-.05.24-.16.35c-.11.09-.23.17-.34.17c-.11 0-.19-.02-.25-.05c-1.28-.69-2.87-1.03-4.75-1.03c-2.05 0-3.88.5-5.5 1.5c-1.34-1-3.17-1.5-5.5-1.5c-1.66 0-3.25.36-4.75 1.07c-.03.01-.07.01-.12.03c-.04.01-.08.02-.13.02c-.11 0-.23-.04-.34-.12a.475.475 0 0 1-.16-.35V6c1.34-1 3.18-1.5 5.5-1.5c2.33 0 4.16.5 5.5 1.5c1.34-1 3.17-1.5 5.5-1.5Z"
      />
    </svg>
  );
}

export function AddBookIcon() {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width="32"
      height="32"
      viewBox="0 0 24 24"
    >
      <path
        fill="currentColor"
        d="m12 18l-4.2 1.8q-1 .425-1.9-.163T5 17.976V5q0-.825.588-1.413T7 3h6v2H7v12.95l5-2.15l5 2.15V11h2v6.975q0 1.075-.9 1.663t-1.9.162L12 18Zm0-13H7h6h-1Zm5 2h-1q-.425 0-.713-.288T15 6q0-.425.288-.713T16 5h1V4q0-.425.288-.713T18 3q.425 0 .713.288T19 4v1h1q.425 0 .713.288T21 6q0 .425-.288.713T20 7h-1v1q0 .425-.288.713T18 9q-.425 0-.713-.288T17 8V7Z"
      />
    </svg>
  );
}

export function BookAddedIcon() {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width="32"
      height="32"
      viewBox="0 0 24 24"
    >
      <path
        fill="currentColor"
        d="m12 18l-4.2 1.8q-1 .425-1.9-.163T5 17.976V5q0-.825.588-1.413T7 3h7q-.5.75-.75 1.438T13 6q0 1.8 1.137 3.175T17 10.9q.575.075 1 .075t1-.075v7.075q0 1.075-.9 1.663t-1.9.162L12 18Zm5.825-11.825L20.65 3.35q.3-.3.713-.3t.712.3q.3.3.3.712t-.3.713L18.525 8.3q-.3.3-.7.3t-.7-.3L15.7 6.875q-.275-.275-.275-.687t.275-.713q.3-.3.713-.3t.712.3l.7.7Z"
      />
    </svg>
  );
}

export function RemoveBookIcon() {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width="32"
      height="32"
      viewBox="0 0 24 24"
    >
      <path
        fill="currentColor"
        d="M16 7q-.425 0-.713-.288T15 6q0-.425.288-.713T16 5h4q.425 0 .713.288T21 6q0 .425-.288.713T20 7h-4Zm-4 11l-4.2 1.8q-1 .425-1.9-.163T5 17.976V5q0-.825.588-1.413T7 3h7q-.5.75-.75 1.438T13 6q0 1.8 1.137 3.175T17 10.9q.575.075 1 .075t1-.075v7.075q0 1.075-.9 1.663t-1.9.162L12 18Z"
      />
    </svg>
  );
}

export function TrashIcon() {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width="32"
      height="32"
      viewBox="0 0 24 24"
    >
      <path
        fill="currentColor"
        d="M7 21q-.825 0-1.413-.588T5 19V6H4V4h5V3h6v1h5v2h-1v13q0 .825-.588 1.413T17 21H7Zm2-4h2V8H9v9Zm4 0h2V8h-2v9Z"
      />
    </svg>
  );
}

export function SearchIcon() {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width="32"
      height="32"
      viewBox="0 0 24 24"
    >
      <path
        fill="currentColor"
        d="M9.5 16q-2.725 0-4.612-1.888T3 9.5q0-2.725 1.888-4.612T9.5 3q2.725 0 4.612 1.888T16 9.5q0 1.1-.35 2.075T14.7 13.3l5.6 5.6q.275.275.275.7t-.275.7q-.275.275-.7.275t-.7-.275l-5.6-5.6q-.75.6-1.725.95T9.5 16Zm0-2q1.875 0 3.188-1.313T14 9.5q0-1.875-1.313-3.188T9.5 5Q7.625 5 6.312 6.313T5 9.5q0 1.875 1.313 3.188T9.5 14Z"
      />
    </svg>
  );
}

export function SadFaceIcon() {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width="32"
      height="32"
      viewBox="0 0 24 24"
    >
      <path
        fill="currentColor"
        d="M12 1.998c5.524 0 10.002 4.478 10.002 10.002c0 5.523-4.478 10-10.002 10c-5.524.001-10-4.477-10-10C1.999 6.476 6.476 1.998 12 1.998Zm0 1.5a8.502 8.502 0 1 0 0 17.003a8.502 8.502 0 0 0 0-17.003Zm0 10c1.631 0 3.16.654 4.281 1.795a.75.75 0 1 1-1.07 1.05a4.485 4.485 0 0 0-3.21-1.346a4.485 4.485 0 0 0-3.217 1.352a.75.75 0 1 1-1.072-1.05A5.984 5.984 0 0 1 12 13.498ZM9 8.75a1.25 1.25 0 1 1 0 2.499A1.25 1.25 0 0 1 9 8.75Zm6 0a1.25 1.25 0 1 1 0 2.499a1.25 1.25 0 0 1 0-2.499Z"
      />
    </svg>
  );
}
