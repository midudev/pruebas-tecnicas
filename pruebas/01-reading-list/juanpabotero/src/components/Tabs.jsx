export function Tabs({
  filteredBooksMapped,
  isReadingListActive,
  setIsReadingListActive,
  readingList,
  clearReadingList
}) {
  return (
    <section
      className="grid grid-cols-2 sm:grid-flow-col sm:grid-cols-none 
      justify-start gap-8 mb-10"
    >
      <button
        className={`font-bold text-white pb-1 border-b ${
          isReadingListActive ? 'border-b-transparent' : ''
        }`}
        onClick={() => setIsReadingListActive(false)}
      >
        Libros disponibles ({filteredBooksMapped.length})
      </button>
      <button
        className={`font-bold text-white pb-1 border-b ${
          isReadingListActive ? '' : 'border-b-transparent'
        }`}
        onClick={() => setIsReadingListActive(true)}
      >
        Lista de lectura ({readingList.length})
      </button>
      <button
        className="text-white text-sm bg-gray-600 px-2 rounded border border-transparent hover:border-gray-400 transition-all duration-200"
        onClick={clearReadingList}
      >
        Limpiar lista
      </button>
    </section>
  );
}
