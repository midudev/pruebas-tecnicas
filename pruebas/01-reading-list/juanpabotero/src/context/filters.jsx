import { createContext, useState } from 'react';

export const FiltersContext = createContext();

export function FiltersProvider({ children }) {
  const [filters, setFilters] = useState({
    genre: 'all',
    author: 'all',
    search: '',
  });

  return (
    <FiltersContext.Provider
      value={{
        filters,
        setFilters,
      }}
    >
      {children}
    </FiltersContext.Provider>
  );
}
