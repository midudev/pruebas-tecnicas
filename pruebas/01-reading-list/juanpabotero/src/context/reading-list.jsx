/* eslint-disable react/prop-types */
import { createContext, useReducer } from 'react';
import {
  readingListInitialState,
  readingListReducer,
} from '../reducers/reading-list.js';

export const ReadingListContext = createContext();

function useReadingListReducer() {
  const [state, dispatch] = useReducer(
    readingListReducer,
    readingListInitialState
  );

  const addToReadingList = (book) =>
    dispatch({
      type: 'ADD_TO_READING_LIST',
      payload: book,
    });

  const removeFromReadingList = (book) =>
    dispatch({
      type: 'REMOVE_FROM_READING_LIST',
      payload: book,
    });

  const clearReadingList = () => dispatch({ type: 'CLEAR_READING_LIST' });

  return { state, addToReadingList, removeFromReadingList, clearReadingList };
}

export function ReadingListProvider({ children }) {
  const { state, addToReadingList, removeFromReadingList, clearReadingList } =
    useReadingListReducer();

  return (
    <ReadingListContext.Provider
      value={{
        readingList: state,
        addToReadingList,
        removeFromReadingList,
        clearReadingList,
      }}
    >
      {children}
    </ReadingListContext.Provider>
  );
}
