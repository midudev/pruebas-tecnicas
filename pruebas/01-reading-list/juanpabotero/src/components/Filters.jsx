import { useId } from 'react';
import { useFilters } from '../hooks/useFilters.js';
import { SearchIcon } from './Icons.jsx';
import '../styles/Filters.css';

export function Filters() {
  const { filters, setFilters } = useFilters();

  const genreFilterId = useId();
  const authorFilterId = useId();
  const searchFilterId = useId();

  const handleChangeGenre = (event) => {
    setFilters((prevState) => ({
      ...prevState,
      genre: event.target.value,
    }));
  };

  const handleChangeAuthor = (event) => {
    setFilters((prevState) => ({
      ...prevState,
      author: event.target.value,
    }));
  };

  const handleChangeSearch = (event) => {
    setFilters((prevState) => ({
      ...prevState,
      search: event.target.value.toLowerCase().trim(),
    }));
  };

  return (
    <section className="flex flex-col sm:flex-row gap-4 sm:gap-12 text-white">
      <div className="flex flex-col gap-2">
        <label htmlFor={genreFilterId}>Filtrar por género:</label>
        <select
          id={genreFilterId}
          onChange={handleChangeGenre}
          className="text-gray-100 bg-gray-600 p-2 rounded-md outline-none focus:ring-1 focus:ring-gray-400"
        >
          <option value="all">Todos</option>
          <option value="Ciencia ficción">Ciencia ficción</option>
          <option value="Fantasía">Fantasía</option>
          <option value="Terror">Terror</option>
          <option value="Zombies">Zombies</option>
        </select>
      </div>

      <div className="flex flex-col gap-2">
        <label htmlFor={authorFilterId}>Filtrar por autor:</label>
        <select
          id={authorFilterId}
          onChange={handleChangeAuthor}
          className="text-gray-100 bg-gray-600 p-2 rounded-md outline-none focus:ring-1 focus:ring-gray-400"
        >
          <option value="all">Todos</option>
          <option value="Bram Stoker">Bram Stoker</option>
          <option value="Douglas Adams">Douglas Adams</option>
          <option value="Frank Herbert">Frank Herbert</option>
          <option value="George Orwell">George Orwell</option>
          <option value="George R. R. Martin">George R. R. Martin</option>
          <option value="H.P. Lovecraft">H.P. Lovecraft</option>
          <option value="J.K. Rowling">J.K. Rowling</option>
          <option value="J.R.R. Tolkien">J.R.R. Tolkien</option>
          <option value="Manel Loreiro">Manel Loreiro</option>
          <option value="Mary Shelley">Mary Shelley</option>
          <option value="Ray Bradbury">Ray Bradbury</option>
          <option value="Stephen King">Stephen King</option>
          <option value="William Gibson">William Gibson</option>
        </select>
      </div>

      <div className="flex flex-col gap-2 sm:max-w-xs w-full relative">
        <label htmlFor={searchFilterId}>Buscar por título:</label>
        <input
          type="text"
          id={searchFilterId}
          placeholder="Dune..."
          onChange={handleChangeSearch}
          className="text-gray-100 bg-gray-600 p-2 rounded-md outline-none focus:ring-1 focus:ring-gray-400"
        />
        <div className='search-icon-container absolute top-10 right-1'>
          <SearchIcon />
        </div>
      </div>
    </section>
  );
}
