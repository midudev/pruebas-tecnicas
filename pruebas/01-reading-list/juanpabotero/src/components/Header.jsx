import { Filters } from './Filters';

export function Header() {
  return (
    <header className="grid gap-8 mb-16">
      <h1 className="text-3xl text-gray-50 font-bold flex gap-3 mb-2">
        Mi Librer√≠a
      </h1>
      <Filters />
    </header>
  );
}
