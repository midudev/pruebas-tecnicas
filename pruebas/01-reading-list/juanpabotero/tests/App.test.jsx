import { render, screen } from '@testing-library/react';
import { describe, expect, test } from 'vitest';
import { Header } from '../src/components/Header';
import { FiltersContext } from '../src/context/filters';
import { library } from '../src/mocks/books.json';

describe('App', () => {
  test('Debe mostrar la cantidad de libros disponibles', () => {
    const filters = {
      genre: 'all',
      author: 'all',
      search: ''
    };
    const initialBooks = library;
    render(
      <FiltersContext.Provider value={{ filters }}>
        <Header initialBooks={initialBooks} filteredBooks={[]} />
      </FiltersContext.Provider>
    );
    const initialBooksCount = screen.getByLabelText('initialBooksCount');
    expect(initialBooksCount).toHaveTextContent('13');
  });
});
