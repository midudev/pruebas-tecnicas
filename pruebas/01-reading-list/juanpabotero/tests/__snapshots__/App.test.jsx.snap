// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > Debe hacer match con el snapshot 1`] = `
<span
  class="font-normal text-gray-50 mt-4"
>
  3
   resultado(s)
</span>
`;
